### PR TITLE
Offchain storage

### LINKED ISSUE #29: Use off-chain storage
### We can use `off-chain indexing` to save NFT's pending queue in `off-chain storage`

1. Replace on-chain `NftPendingQueue` with off-chain storage
2. Add mutex lock's to avoid duplicates

### PR DESCRIPTION
**Pull request by korzewski**
Closes #29

### CODE REVIEW

**Comment by korzewski** [comment]
This PR contains some cleanup like files splitting.

Quick overview what happen with NFT flow:
- User call `nft_request()`, this add it to `PendingNftQueue` (on-chain)
- `on_finalize()` copy `PendingNftQueue` to offchain storage `new_nft_requests`. Then clean up onchain `PendingNftQueue` so it's never added to block. I'm using it only as communication layer
- From here everything is offchain storage.` offchain hook_init()` is executed after each block. First we update `nft_pending_queue` by pushing all `new_nft_requests`
- Now `nft_pending_queue` is up to date and we can use it as before
- Main purpose of `pending queue` in offchain worker is the case of additional http requests or other heavy computations

This may be useful for all metadata storage. On-chain would keep only `key_hash` which is offchain storage database index key where we can keep more data without being worry of on-chain storage size.

I've decided to do not write more unit tests at this stage. We may need to rewrite it slightly when final direction is made.

**Review by PierreOssun — DISMISSED** [review]
Good for me

File: pallets/nft/src/lib.rs [hunk 1]
```diff
@@ -168,116 +153,57 @@ pub mod pallet {
...
-			}
+		fn on_finalize(block_number: T::BlockNumber) {
+			let key = offchain::offchain_new_nft_requests_key();
+			let pending_nft_queue = PendingNftQueue::<T>::get();
+
+			sp_io::offchain_index::set(&key.0, &pending_nft_queue.encode());
+			
+			debug::info!("--- on_finalize block_number: {:?}, new_nft_requests_key: {:x}, new_nft_requests_value: {:?}", block_number, key, pending_nft_queue);
+			
+			PendingNftQueue::<T>::put(PendingNftQueueOf::<T>::new());
```
Comment (aelesbao) [inline pallets/nft/src/lib.rs]: Any reason why you didn't use `PendingNftQueue::<T>::mutate()` here instead?

File: pallets/nft/src/lib.rs [hunk 2]
```diff
@@ -168,116 +153,57 @@ pub mod pallet {
...
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
-		fn offchain_worker(_block_number: T::BlockNumber) {
-			let pending_nft_queue = NftPendingQueue::<T>::get();
-			if pending_nft_queue.len() > 0 {
-				Self::execute_nft_from_pending_queue(pending_nft_queue[0].clone());
-			}
+		fn on_finalize(block_number: T::BlockNumber) {
+			let key = offchain::offchain_new_nft_requests_key();
+			let pending_nft_queue = PendingNftQueue::<T>::get();
```
Comment (aelesbao) [inline pallets/nft/src/lib.rs]: What happens here on the first access when there's no value stored?

File: pallets/nft/src/local_storage.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            Ok(Ok(x)) => Ok(x),
+            Err(e) => Err(e),
+            Ok(Err(_x)) => Err(Error::<C>::OffchainLock),
+        }
+    }
+}
+
+pub fn get_nft_key_hash<T: Config>(class_id: T::ClassId, token_data: T::TokenData) -> KeyHash {
+    (class_id, token_data).using_encoded(blake2_128)
+}
```
Comment (aelesbao) [inline pallets/nft/src/local_storage.rs]: I'm not sure if we added it to the coding guidelines, but it's [always a good idea to leave a newline at the end of the file](https://unix.stackexchange.com/a/18789/370378). 😃

File: pallets/nft/src/offchain.rs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        Ok(_) => {
+            if let Ok(queue) = pending_nft_queue.get::<T>() {
+                if queue.len() > 0 {
+                    debug::info!("--- Pending nft queue key: {:x}, value: {:?}", VecKey(NFT_PENDING_QUEUE.to_vec()), queue);
+
+                    let pending_nft = &queue[0];
+                    execute_nft_from_pending_queue::<T>(pending_nft.clone());
+
+                    let _: Result<_, Error<T>> = pending_nft_queue.mutate(|x| remove_vector_item(x, pending_nft));
+                }
```
Comment (aelesbao) [inline pallets/nft/src/offchain.rs]: Why don't you loop through the queue here?

File: pallets/nft/src/offchain.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            return Err(Error::<T>::OffchainSignedTxError)
+        }
+
+        debug::info!("--- Send signed - Ok");
+        return Ok(());
+    } 
+
+    debug::error!("--- Send signed - No local account available");
+    return Err(Error::<T>::NoLocalAccountForSigning);
+}
```
Comment (aelesbao) [inline pallets/nft/src/offchain.rs]: Same here about a newline at the EOF.

**Review by aelesbao — CHANGES_REQUESTED** [review]

File: pallets/nft/src/local_storage.rs (hunk 3 above)
Comment (nblogist) [reply pallets/nft/src/local_storage.rs]: Thank you! Coding guidelines have been updated

**Review by nblogist — CHANGES_REQUESTED** [review]
After other reviews are finalized and #34 is merged, please merge `ibtida` to your branch.

File: pallets/nft/src/lib.rs (hunk 1 above)
Comment (korzewski) [reply pallets/nft/src/lib.rs]: Good point

File: pallets/nft/src/lib.rs (hunk 2 above)
Comment (korzewski) [reply pallets/nft/src/lib.rs]: It will get a default value which is empty `Vec` of type `PendingNft`

**Review by nblogist — DISMISSED** [review]

**Review by aelesbao — DISMISSED** [review]
LGTM

**Comment by korzewski** [comment]
All was accepted before `ibtida` merge. Should be good to go now

**Review by aelesbao — APPROVED** [review]

**Review by nblogist — APPROVED** [review]
Looks good

**Pull request merged**